scalr/data_source_scalr_module_versions.go

```diff
@@ -0,0 +1,124 @@
+package scalr
+
+import (
+	"context"
+	"errors"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/scalr/go-scalr"
+)
+
+func dataSourceModuleVersions() *schema.Resource {
+	return &schema.Resource{
+		Description: "Retrieves a list of module versions by module source or module id.",
+		ReadContext: dataSourceModuleVersionsRead,
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Description:  "The identifier of а module. Example: `mod-xxxx`",
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				AtLeastOneOf: []string{"id", "source"},
+			},
+			"source": {
+				Description: "The source of a module.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+			},
+			"versions": {
+				Description: "The list of semantic versions.",
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Description: "The identifier of а module version. Example: `modver-xxxx`",
+							Type:        schema.TypeString,
+							Required:    true,
+						},
+						"version": {
+							Description: "The semantic version. Example: `1.2.3`.",
+							Type:        schema.TypeString,
+							Required:    true,
+						},
+						"status": {
+							Description: "The status of a module version. Possible values: `ok`, `pending`, `not_uploaded`, `errored`",
+							Type:        schema.TypeString,
+							Required:    true,
+						},
+					},
+				},
+			},
+		}}
+}
+
+func dataSourceModuleVersionsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	scalrClient := meta.(*scalr.Client)
+
+	moduleID := d.Get("id").(string)
+	moduleSource := d.Get("source").(string)
+
+	var module *scalr.Module
+
+	if moduleID != "" {
+		log.Printf("[DEBUG] Read module with ID: %s", moduleID)
+		var err error
+		module, err = scalrClient.Modules.Read(ctx, moduleID)
+		if err != nil {
+			if errors.Is(err, scalr.ErrResourceNotFound) {
+				return diag.Errorf("Could not find module with ID '%s'", moduleID)
+			}
+			return diag.Errorf("Error retrieving module: %v", err)
+		}
+
+		if moduleSource == "" {
+			_ = d.Set("source", module.Source)
+		} else if module.Source != moduleSource {
+			return diag.Errorf("Could not find module with ID '%s' and source '%s'", moduleID, moduleSource)
+		}
+
+	} else if moduleSource != "" {
+		log.Printf("[DEBUG] Read module with source: %s", moduleSource)
+		var err error
+		module, err = scalrClient.Modules.ReadBySource(ctx, moduleSource)
+		if err != nil {
+			if errors.Is(err, scalr.ErrResourceNotFound) {
+				return diag.Errorf("Could not find module with source '%s'", moduleSource)
+			}
+			return diag.Errorf("Error retrieving module: %v", err)
+		}
+
+	} else {
+		return diag.Errorf("Error retrieving module: either 'id' or 'source' is required")
```

Comment:
Won't happen — it is guaranteed by `AtLeastOneOf` schema validator.

Reply:
AtLeastOneOf is declared in the Schema, but in the function this case is implicit, so in this case it's better to cover this case with an error.

Reply:
Should we duplicate [all](https://github.com/Scalr/terraform-provider-scalr/blob/a92bb6bda8a9f532a359ea1f7de6b6a783dbc087/scalr/resource_scalr_workspace.go#L70-L71) [declarative](https://github.com/Scalr/terraform-provider-scalr/blob/a92bb6bda8a9f532a359ea1f7de6b6a783dbc087/scalr/resource_scalr_provider_configuration.go#L79) [validations](https://github.com/Scalr/terraform-provider-scalr/blob/a92bb6bda8a9f532a359ea1f7de6b6a783dbc087/scalr/resource_scalr_service_account.go#L47) in code?

Reply:
It depends on the logic:
 - when we collect all attributes and pass them in filter when use api client, we shouldn't check the case when none of required attributes are set;
 - in my case I have two cases for each required attribute and we use `if/else if`, and it's a good practice to add extra `else` to cover all cases.